cardano-tx-generator/src/Cardano/Benchmarking/GeneratorTx/Callback.hs

```diff
@@ -0,0 +1,100 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+module Cardano.Benchmarking.GeneratorTx.Callback
+  (
+    mkCallback
+  ) where
+
+import           Prelude (error)
+import           Cardano.Prelude hiding (TypeError, show)
+
+import qualified Ouroboros.Consensus.Cardano as Consensus
+import           Ouroboros.Consensus.Config
+                   ( configBlock, configCodec)
+import           Ouroboros.Consensus.Config.SupportsNode
+                   (ConfigSupportsNode(..), getNetworkMagic)
+import           Ouroboros.Consensus.Node.ProtocolInfo
+                   (ProtocolInfo (..))
+import           Ouroboros.Network.NodeToClient (IOManager)
+
+import           Cardano.Chain.Slotting
+import           Cardano.Api
+import           Cardano.Api.Typed
+import qualified Cardano.Api.Typed as Api
+
+-- Node imports
+import           Cardano.Node.Types (SocketPath(..))
+import           Cardano.Tracing.OrphanInstances.Byron()
+import           Cardano.Tracing.OrphanInstances.Common()
+import           Cardano.Tracing.OrphanInstances.Consensus()
+import           Cardano.Tracing.OrphanInstances.Network()
+import           Cardano.Tracing.OrphanInstances.Shelley()
+
+import Cardano.Benchmarking.GeneratorTx.Benchmark
+import Cardano.Benchmarking.GeneratorTx.Era
+import Cardano.Benchmarking.GeneratorTx.NodeToNode
+import Cardano.Benchmarking.GeneratorTx
+import Cardano.Benchmarking.GeneratorTx.Genesis (GeneratorFunds)
+
+type Funding era = Benchmark -> GeneratorFunds -> ExceptT TxGenError IO (SigningKey PaymentKey, [(TxIn, TxOut era)])
+type BenchmarkAction era
+      = Benchmark -> (SigningKey PaymentKey, [(TxIn, TxOut era)]) -> ExceptT TxGenError IO ()
+
+type Action era = Benchmark -> GeneratorFunds -> ExceptT TxGenError IO ()
+
+mkCallback
+  :: forall blok ptcl era.
+     (
+       IsShelleyBasedEra era
+     , ConfigSupportsTxGen CardanoMode era
+     )
+  =>  Consensus.Protocol IO blok ptcl
+  -> Maybe NetworkMagic
+  -> Bool
+  -> IOManager
+  -> SocketPath
+  -> BenchTracers IO CardanoBlock
+  -> Proxy era
+  -> Action era
```

Comment:
Btw, also, I'm a bit confused by the name -`mkCallback` is _very_ generic..

Reply:
Yes I'm not really happy with name either.
This is the replacement for `mkMode`.
Instead of wrapping all the Arguments into a `Mode` type it applies then
directly to the `secureFunds` run `runBenchmark` functions.
Except for the `era` which is provided later.
It returns and Action which is then run later in `Run.hs`.
However, we also could run the Action here right away.
Or, the other way around, inline this in `Run.hs`

Reply:
I think it would be worth resolving these questions before we merge.

Reply:
Its not a good name because the structure itself is not good.
The next ticket is to add support for benchmarking across era boundaries,
which probably require a rewrite of this part.
But yes 'm going to rename the module and the function.

Reply:
I've renamed the module to `LocalProtocolDefinition` in the next rewrite I want to move everything that deals with
loading and parsing the protocol definition from `Run.hs` to this module.